### PR TITLE
Add new Tasks.Feed task: CreateInternalBlobFeed

### PR DESCRIPTION
**Pull request by JohnTortugo**
Closes: https://github.com/dotnet/arcade/issues/3218

This task will be used temporarily to create an Azure Storage folder + initialize it with a NuGet feed. This is part of a workaround to publish to internal feeds.

Some tests of this task created this container + folder: cesarfeed / dotnet-feed-internal in _HelixStaging_ subscription.

### CODE REVIEW

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        [Required]
+        public string AzureDevOpsFeedsBaseUrl { get; set; }
+
+        [Required]
+        public string AzureStorageAccountName { get; set; }
+
+        [Required]
+        public string AzureStorageAccountKey { get; set; }
+
+        private const string baseUrlRegex = @"https:\/\/(?<containername>[^\.]+).*";
```
Comment (riarenas) [inline src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: >https:\/\/(?<containername>[^\.]+).* [](start = 46, length = 36)

discussed offline, but this regex needs to be tweaked a bit to actually capture the right container name.

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                        needsUniqueName = false;
+                    }
+                } while (needsUniqueName);
+
+                // Initialize the feed using sleet
+                SleetSource sleetSource = new SleetSource()
+                {
+                    Name = baseFeedName,
+                    Type = "azure",
+                    BaseUri = $"{AzureDevOpsFeedsBaseUrl}/{containerName}/{baseFeedName}",
```
Comment (riarenas) [inline src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: >{AzureDevOpsFeedsBaseUrl}/{containerName} [](start = 32, length = 41)

The baseURL is already scoped to the container. Do we need to add it again?

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                    AccountName = AzureStorageAccountName,
+                    Container = containerName,
+                    FeedSubPath = $"{baseFeedName}",
```
Comment (riarenas) [inline src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: I believe @jcagme mentioned that this one needed to include the proxy URI as well?

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/common/AzureStorageUtils.cs [hunk 4]
```diff
@@ -134,9 +134,9 @@ public string CreateSASToken(int tokenExpirationInDays, SharedAccessBlobPermissi
             return Container.GetSharedAccessSignature(sasConstraints);
         }
 
-        public Task<bool> CheckIfContainerExistsAsync()
```
Comment (riarenas) [inline src/Microsoft.DotNet.Build.Tasks.Feed/src/common/AzureStorageUtils.cs]: This wasn't used anywhere?

**Comment by riarenas** [comment]
LGTM with some comments. But I think @jcagme needs to approve before merging to make sure the sleet configuration looks right.

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs (hunk 3 above)
Comment (jcagme) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: No, this is ok, the one which needs the proxy is BaseUri

---
In reply to: [300150405](https://github.com/dotnet/arcade/pull/3224#discussion_r300150405) [](ancestors = 300150405)

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+                    }
+                } while (needsUniqueName);
+
+                // Initialize the feed using sleet
+                SleetSource sleetSource = new SleetSource()
+                {
+                    Name = baseFeedName,
+                    Type = "azure",
+                    BaseUri = $"{AzureDevOpsFeedsBaseUrl}/{containerName}/{baseFeedName}",
+                    AccountName = AzureStorageAccountName,
```
Comment (jcagme) [inline src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: >AccountName = AzureStorageAccountName [](start = 20, length = 37)

This shouldn't be needed

**Comment by riarenas** [comment]
Do you want to make the switch to use the new task in the proj files as part of this PR too?

**Comment by JohnTortugo** [comment]
> Do you want to make the switch to use the new task in the proj files as part of this PR too?

I'll make a separate PR. I have to get a new Tasks.Feed version and put it in the SDK anyway..

**Comment by jcagme** [comment]
We'll also need modifications/new classes to publish to the feed. I'll send you the code

**Comment by riarenas** [comment]
Sounds good.

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs (hunk 5 above)
Comment (JohnTortugo) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: Looks like Sleet requires it: {"Value cannot be null.\r\nParameter name: accountName"}

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/common/AzureStorageUtils.cs (hunk 4 above)
Comment (JohnTortugo) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/common/AzureStorageUtils.cs]: Nope.

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs (hunk 1 above)
Comment (JohnTortugo) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: Fixed. Going up in next push

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs (hunk 5 above)
Comment (jcagme) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: That's so weird. My local repro doesn't need that

---
In reply to: [300157597](https://github.com/dotnet/arcade/pull/3224#discussion_r300157597) [](ancestors = 300157597)

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs (hunk 2 above)
Comment (JohnTortugo) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: Yeah, we don't.

File: src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs (hunk 5 above)
Comment (JohnTortugo) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: This is required by the check I added on line 450 of BlobFeedAction.cs . I can remove the check but I think the benefit of keeping it are greater than not having the parameter.

Comment (jcagme) [reply src/Microsoft.DotNet.Build.Tasks.Feed/src/CreateInternalBlobFeed.cs]: Makes sense

**Pull request merged**